### PR TITLE
Extended tid chapter for m mode capabilities

### LINKED ISSUE #283: The TID extension is insufficient for the current compartmentalisation models
The TID extension defines two registers: STID(C) and UTID(C) of which the latter is read-only. In order to support compartmentalisation models on multiple privilege levels running independently, I propose to add a read-write CSR on all support privilege levels resulting in: MTID(C), STID(C), and UTID(C).

Furthermore, I propose enforcing the following permissions on the CSRs:
MTID(C) is writeable with from M mode with ASR permission, but readable from M mode without ASR permission.
STID(C) is writeable with from M mode with ASR permission or S mode with ASR permission, but readable from M mode without ASR permission or S mode without ASR permission.
UTID(C) is writeable with from M mode with ASR permission or S mode with ASR permission or U mode with ASR permission, but readable from M mode without ASR permission or S mode without ASR permission or U mode without ASR permission.

I propose the following indexes for the CSRs:

`mtidc`: 0x780
`stidc`: 0x580 (unchanged)
`utidc`: 0x480

### PR DESCRIPTION
**Pull request by francislaus**
This will close #283.

### CODE REVIEW

File: src/csv/CHERI_CSR.csv [hunk 1]
```diff
@@ -30,5 +30,6 @@ direct write if address didn't change","✔","","","","Zcmt","Jump Vector Table
...
-"utidc","0xca0","utid","U","URO","tag=0, otherwise undefined","N/A","N/A","","","","✔","Zstid","Read-only copy of stidc","","","","","","","","","","","","","","","","","","","","",""
-"stidc","0x580","stid","S","SRW, <<asr_perm>>","tag=0, otherwise undefined","Update the CSR using <<SCADDR>>.","direct write","","","","✔","Zstid","Secure thread ID","","","","","","","","","","","","","","","","","","","","",""
+"utidc","0x480","utid","U","Read: U, Write: U, <<asr_perm>>","tag=0, otherwise undefined","Update the CSR using <<SCADDR>>.","direct write","","","","✔","Zstid","User thread ID","","","","","","","","","","","","","","","","","","","","",""
+"stidc","0x580","stid","S","Read: S, Write: S, <<asr_perm>>","tag=0, otherwise undefined","Update the CSR using <<SCADDR>>.","direct write","","","","✔","Zstid","Supervisor thread ID","","","","","","","","","","","","","","","","","","","","",""
+"mtidc","0x780","mtid","M","Read: M, Write: M, <<asr_perm>>","tag=0, otherwise undefined","Update the CSR using <<SCADDR>>.","direct write","","","","✔","Zstid","Machine thread ID","","","","","","","","","","","","","","","","","","","","",""
```
Comment (arichardson) [inline src/csv/CHERI_CSR.csv:35]: Will need to update the source sheet as well.

File: src/tid-ext.adoc [hunk 2]
```diff
@@ -69,10 +103,11 @@ include::img/stidcreg.edn[]
 [#utidc,reftext="utidc"]
 ==== User Thread Identifier Capability (utidc)
 
-The <<utidc>> register is an CLEN-wide read-only capability register.
-It is the capability extension of the <<utid>> register. <<utidc>> is
-a read-only copy of the <<stidc>> register. On reset the tag of <<utidc>>
-will be set to 0 and the remainder of the data is UNSPECIFIED.
+The <<utidc>> register is an CLEN-bit read-write capability register.
+It is the capability extension of the <<utid>> register.
+It is used to identify the current thread in the supervisor mode.
```
Comment (arichardson) [inline src/tid-ext.adoc:108]:
```suggestion
It is used to identify the current thread in the user mode.
```

File: src/tid-ext.adoc [hunk 3]
```diff
@@ -44,23 +65,36 @@ include::img/stidreg.edn[]
...
 
 When {cheri_base_ext_name} is implemented, the {tid_ext_name} CSRs are
 extended as follows:
 
+[#mtidc,reftext="mtidc"]
+==== Machine Thread Identifier Capability (mtidc)
+
+The <<mtidc>> register is an CLEN-bit read-write capability register.
+It is the capability extension of the <<mtid>> register.
+It is used to identify the current thread in the supervisor mode.
```
Comment (arichardson) [inline src/tid-ext.adoc:83]:
```suggestion
It is used to identify the current thread in the machine mode.
```

File: src/csv/CHERI_CSR.csv (hunk 1 above)
Comment (francislaus) [reply src/csv/CHERI_CSR.csv:35]: That is pulled from the source sheet (https://docs.google.com/spreadsheets/d/1nyxKamsYZaeTyG8qP-JX4_oQwcuQ_4nZ_Ihm3RK0NEY/edit#gid=1927549494). Did you mean that?

File: src/csv/CHERI_CSR.csv [hunk 4]
```diff
@@ -30,5 +30,6 @@ direct write if address didn't change","✔","","","","Zcmt","Jump Vector Table
...
 "ddc","0x416","","U","URW","<<infinite-cap>>","","","","✔","","","{cheri_default_ext_name}","User Default Data Capability","","","","","","","","","","","","","","","","","","","","",""
 "dinfc","0x7bd","","D","DRW","<<infinite-cap>>","","","","","","✔","Sdext","Source of <<infinite-cap>> capability in debug mode, writes are ignored","","","","","","","","","","","","","","","","","","","","",""
-"utidc","0xca0","utid","U","URO","tag=0, otherwise undefined","N/A","N/A","","","","✔","Zstid","Read-only copy of stidc","","","","","","","","","","","","","","","","","","","","",""
-"stidc","0x580","stid","S","SRW, <<asr_perm>>","tag=0, otherwise undefined","Update the CSR using <<SCADDR>>.","direct write","","","","✔","Zstid","Secure thread ID","","","","","","","","","","","","","","","","","","","","",""
+"utidc","0x480","utid","U","Read: U, Write: U, <<asr_perm>>","tag=0, otherwise undefined","Update the CSR using <<SCADDR>>.","direct write","","","","✔","Zstid","User thread ID","","","","","","","","","","","","","","","","","","","","",""
```
Comment (tariqkurd-repo) [inline src/csv/CHERI_CSR.csv:33]: 0x480 conflicts with mcycleh, so this needs to be 0xca0

Comment (tariqkurd-repo) [inline src/csv/CHERI_CSR.csv:33]: I think this is the only case of a U-mode register needing ASR? Should we change the description of ASR as now we say it's for _privileged_ register access?

File: src/tid-ext.adoc [hunk 5]
```diff
@@ -1,41 +1,62 @@
...
-=== Supervisor-Level and Unprivileged CSRs
+
+
+=== Machine-Level, Supervisor-Level and Unprivileged CSRs
+
+[#mtid,reftext="mtid"]
+==== Machine Thread Identifier (mtid)
+
+The <<mtid>> register is an MXLEN-bit read-write register. It is used to
+identify the current thread in the machine mode. The reset value of this
```
Comment (tariqkurd-repo) [inline src/tid-ext.adoc:48]:
```suggestion
identify the current thread in machine mode. The reset value of this
```

File: src/tid-ext.adoc [hunk 6]
```diff
@@ -1,41 +1,62 @@
...
+.Supervisor thread identifier register
+include::img/mtidreg.edn[]
 
 [#stid,reftext="stid"]
 ==== Supervisor Thread Identifier (stid)
 
 The <<stid>> register is an SXLEN-bit read-write register. It is used to
-identify the current thread. The reset value of this register is
-UNSPECIFIED.
+identify the current thread in the supervisor mode. The reset value of
```
Comment (tariqkurd-repo) [inline src/tid-ext.adoc:58]:
```suggestion
identify the current thread in supervisor mode. The reset value of
```

File: src/tid-ext.adoc [hunk 7]
```diff
@@ -44,23 +65,36 @@ include::img/stidreg.edn[]
 [#utid,reftext="utid"]
 ==== User Thread Identifier (utid)
 
-The <<utid>> register is an UXLEN-bit read-only register. It is a read-only
-copy of the <<stid>> register. The reset value of this register is
-UNSPECIFIED.
+The <<utid>> register is an UXLEN-bit read-write register. It is used to
+identify the current thread in the user mode. The reset value of
```
Comment (tariqkurd-repo) [inline src/tid-ext.adoc:69]:
```suggestion
identify the current thread in user mode. The reset value of
```

File: src/csv/CHERI_CSR.csv (hunk 4 above)
Comment (francislaus) [reply src/csv/CHERI_CSR.csv:33]: I think it is the only CSR that needs ASR permission. This is how ASR is described now: https://github.com/riscv/riscv-cheri/blob/9661d31152451f0aa406ad74ecbc4959cb5d06e5/src/cap-description.adoc?plain=1#L121-L122

Did you mean this one or something else? @tariqkurd-repo

Comment (tariqkurd-repo) [reply src/csv/CHERI_CSR.csv:33]: yes - that one - util[c] isn't _privileged_ so we'd need to update the definition - assuming that _privileged_ means M-mode / S-mode accessible.

Comment (PRugg-Cap) [reply src/csv/CHERI_CSR.csv:33]: For what it's worth, I think we were always planning to exempt `mcycle`, `minstret` etc from ASR checks as well. I don't know to what extend we need that, or whether they're redundant with the `cycle`, `instret` "unprivileged" registers. @jrtc27 ?

**Review by arichardson — APPROVED** [review]
This LGTM although we may want to call out that we have one user-only use of ASR (which means the code to write it needs almighty permissions on RV32).

File: src/csv/CHERI_CSR.csv (hunk 4 above)
Comment (tariqkurd-repo) [reply src/csv/CHERI_CSR.csv:33]: maybe it's time to update the definition then?

Comment (tariqkurd-repo) [reply src/csv/CHERI_CSR.csv:33]: @francislaus can we change

`Access System Registers Permission (ASR):: Allow access to privileged CSRs.`

to 

`Access System Registers Permission (ASR):: Allow access to privileged CSRs and <<ustid>>, <<ustidc>>.`

and I think we need another issue to review ASR more generally.
but at least this makes the ASR statement correct.

Comment (tariqkurd-repo) [reply src/csv/CHERI_CSR.csv:33]: The following description matches my understanding of this PR:

```
Access System Registers Permission (ASR):: Allow read and write access to all privileged (M-mode and S-mode) CSRs with the following exceptions:
. <<utid>>, <<utidc>>, <<stid>>, <<stidc>>, <<mtid>>, <<mtidc>> all require ASR access for writing and not for reading
``

**Pull request merged**